src/OpenTelemetry.Instrumentation.AWSLambda/Implementation/AWSMessagingUtils.cs

```diff
@@ -0,0 +1,149 @@
+using Newtonsoft.Json;
+using OpenTelemetry.Context.Propagation;
+
+namespace OpenTelemetry.Instrumentation.AWSLambda.Implementation;
+
+internal class AWSMessagingUtils
+{
+    // SNS attribute types: https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html
+    private const string SnsAttributeTypeString = "String";
+    private const string SnsAttributeTypeStringArray = "String.Array";
+    private const string SnsMessageAttributes = "MessageAttributes";
+
+    internal static PropagationContext ExtractParentContext(SQSEvent sqsEvent)
+    {
+        if (sqsEvent == null)
+        {
+            return default;
+        }
+
+        // We assume there can be only one parent that's why we consider only a single (the last) record as the carrier.
+        var message = sqsEvent.Records.LastOrDefault();
+        return ExtractParentContext(message);
+    }
+
+    internal static PropagationContext ExtractParentContext(SQSEvent.SQSMessage sqsMessage)
+    {
+        if (sqsMessage == null)
+        {
+            return default;
+        }
+
+        var parentContext = Propagators.DefaultTextMapPropagator.Extract(default, sqsMessage.MessageAttributes, SqsMessageAttributeGetter);
+        if (parentContext == default)
+        {
+            // SQS subscribed to SNS topic with raw delivery disabled case, i.e. SNS record serialized into SQS body.
+            // https://docs.aws.amazon.com/sns/latest/dg/sns-large-payload-raw-message-delivery.html
+            SNSEvent.SNSMessage snsMessage = GetSnsMessage(sqsMessage);
+            if (snsMessage != null)
+            {
+                parentContext = ExtractParentContext(snsMessage);
+            }
+        }
+
+        return parentContext;
+    }
+
+    internal static PropagationContext ExtractParentContext(SNSEvent snsEvent)
+    {
+        if (snsEvent == null)
+        {
+            return default;
+        }
+
+        // We assume there can be only one parent that's why we consider only a single (the last) record as the carrier.
+        var record = snsEvent.Records.LastOrDefault();
+        return ExtractParentContext(record);
+    }
+
+    internal static PropagationContext ExtractParentContext(SNSEvent.SNSRecord record)
+    {
+        return (record?.Sns != null) ?
+            Propagators.DefaultTextMapPropagator.Extract(default, record.Sns.MessageAttributes, SnsMessageAttributeGetter) :
+            default;
+    }
+
+    internal static PropagationContext ExtractParentContext(SNSEvent.SNSMessage message)
+    {
+        return (message != null) ?
+            Propagators.DefaultTextMapPropagator.Extract(default, message.MessageAttributes, SnsMessageAttributeGetter) :
+            default;
+    }
+
+    private static IEnumerable<string> SqsMessageAttributeGetter(IDictionary<string, SQSEvent.MessageAttribute> attributes, string attributeName)
+    {
+        SQSEvent.MessageAttribute attribute = attributes.GetValueByKeyIgnoringCase(attributeName);
```

Comment:
I don't think we should need to ignore casing here, it's not HTTP. Isn't it rather expensive as well? Same for SNS.